src/universe/visualization/img_tools.py

```diff
@@ -14,3 +19,352 @@ def rotate(img: np.ndarray, angle: float) -> np.ndarray:
+
+        :param title: title of the plot
+        :param axis: whether to show the axis or not
+        """
+
+        # add legend to plot
+        legend = []
+        for layer in self.layers:
+            if layer.name:
+                legend.append(mpatches.Patch(color=[x / 255 for x in layer.color], label=layer.name))
+        if legend:
+            plt.legend(handles=legend, loc='lower right')
+
+        if title:
+            plt.title(title)
+        if not axis:
+            plt.axis('off')
+        # show plot
+        plt.imshow(self.image)
+
+    def save(self, path: str) -> bool:
+        """
+        Saves the image with the layers drawn on it to the given path.
+
+        :param path: path to save the image to
+        :return: whether the image was saved successfully or not
+        """
+        return cv.imwrite(path, cv.cvtColor(self.image, cv.COLOR_RGB2BGR))
+
+    def resize(self, width: Optional[int] = None, height: Optional[int] = None) -> None:
+        """
+        Resizes the image to the given width and height.
+
+        :param width: new width of the image
+        :param height: new height of the image
+        """
+        if width:
+            if width <= 0:
+                raise ValueError('Width must be greater than 0')
+            else:
+                self.width = width
+        if height:
+            if height <= 0:
+                raise ValueError('Height must be greater than 0')
+            else:
+                self.height = height
+
+    def scale(self, multiplier: float) -> None:
+        """
+        Scales the image by the given multiplier.
+
+        :param multiplier: multiplier to scale the image by
+        """
+        if multiplier <= 0:
+            raise ValueError('Multiplier must be greater than 0')
+
+        self.width = round(self.width * multiplier)
+        self.height = round(self.height * multiplier)
+
+
+def grayscale_image(image: np.ndarray) -> np.ndarray:
+    """
+    Converts the given image to grayscale and then to RGB color space.
+
+    :param image: image to convert (RGB or grayscale)
+    :return: grayscale image converted to RGB color space
+    """
+    if image is None:
+        raise ValueError('Image does not exist')
+
+    # if the image is not grayscale, convert it to grayscale
+    if len(image.shape) == 3:
+        image = cv.cvtColor(image, cv.COLOR_RGB2GRAY)
+
+    # convert the image to RGB color space
+    return cv.cvtColor(image, cv.COLOR_GRAY2RGB)
+
+
+def darken_image(image: np.ndarray, darkness: float = 0.5) -> np.ndarray:
+    """
+    Darkens the given image.
+
+    :param image: image to darken
+    :param darkness: percentage of darkness (0-100)
+    :return: darkened image
+    """
+    if image is None:
+        raise ValueError('Image does not exist')
+    if not (0. <= darkness <= 1.):
+        print('Darkness percentage must be between 0 and 1')
+        darkness = min(max(0., darkness), 1.)
```

Comment:
We should only check these conditions once. Now, it is checked in this method, and also checked in the layer class' implementation. 
This method is public, so if we want to enforce these values we must check the condition here. The layer class' methods are also public, but they are using these methods, therefore we could design the class without additional checks.